### PR TITLE
[TESTNET-14] Dydx deposit flow

### PR DESCRIPTION
**Pull request by Akhilleshgoswami**
This PR has code for operations   that will be perform on Cruize and Dydx.

1.  code in the PR  -
 -  apis  for loan to  borrow , repay ,  from Aave.
 -   load the smart contract.
 -  handle exceptions ,
 -  refactor the code  .
 -  handling transactions and build transactions.
 
2.  changes in this PR - 
-   folder  name = dydx_withdrawal to dydx_operations.
-   class name =. dydx_withdrawal to dydx_operations.
-   urls.py 

      
4. Remaining task  - 
 -  testing the function for the mainnet 
 -  testing dydx deposit
 -  loading contracts with their address and testing them.

### CODE REVIEW

File: cruize_operations/views.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+from rest_framework import status
+from rest_framework.response import Response
+from rest_framework.viewsets import GenericViewSet
+from cruize_operations import AmountSerializers
+from services.contracts.cruize.cruize_contract import Cruize
+
+
+class CruizeOperations(GenericViewSet):
+    def borrow(self, request):
+        self.serializer_class = AmountSerializers
```
Comment (Akhilleshgoswami) [inline cruize_operations/views.py]: no need of this

File: dydx_operations/views.py [hunk 2]
```diff
@@ -60,16 +62,27 @@ def fast_withdrawal(self, request):
             result["error"] = str(e)
             return Response(result, status.HTTP_500_INTERNAL_SERVER_ERROR)
 
+    def deposit(self, request):
```
Comment (Akhilleshgoswami) [inline dydx_operations/views.py:65]: no need of this

File: services/contracts/cruize/cruize_contract.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            tnx = self.transaction_manager.sign_transactions(contract_transaction)
+            result["transaction_hash"] = tnx
+            return result
+        except ValueError as e:
+            return contract_exceptions.contract_exceptions(e)
+        except Exception as e:
+            result["error"] = e
+            return result
+
+    def borrow_from_aave(self, amount):
```
Comment (Akhilleshgoswami) [inline services/contracts/cruize/cruize_contract.py:46]: no need of this

File: services/contracts/starkex_contracts/starkex_contract.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            return signed_tx
+
+        except ValueError as e:
+            return contract_exceptions.contract_exceptions(e)
+
+        except Exception as e:
+            result["error"] = e
+            return result
+
+    def deposit(self, amount):
```
Comment (Akhilleshgoswami) [inline services/contracts/starkex_contracts/starkex_contract.py]: no need of this

File: services/contracts/cruize/cruize_contract.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    def __init__(self):
+        self.load_contract = LoadContracts()
+        self.transaction_manager = TransactionManager()
+        contract_abi = open("services/contracts/contract_abis/Starkware.json")
+        self.contract = self.load_contract.load_contracts(
+            config("STARK_EX_CONTRACT"), contract_abi
+        )
+        self.w3 = self.load_contract.web3_provider()
+
+    # this function will deposit to cruzie contract from our  curize wallet .
```
Comment (Prithviraj8) [inline services/contracts/cruize/cruize_contract.py]: cruzie -> cruize
curize -> cruize

File: utilities/constant.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+AAVE_MAINNET_CONTRACT_ADDRESS = "AAVE"
+CRUIZE_CONTRACT_ADDRESS = "CRUIZE"
+WALLET_ADDRESS = "0xE0E24a32A7e50Ea1c7881c54bfC1934e9b50B520"
+USDC_ADDRESS = ""
+LINK_ADDRESS = "0x01BE23585060835E02B77ef475b0Cc51aA1e0709"
+WALLET_ADDRESS_2 = "0xBD43b05ac20F96B808d3c644c3CF7add86ad5F58"
+
+# price params
+MAX_WAIT_SECONDS = 60
+SAMPLE_SIZE = 20
```
Comment (Prithviraj8) [inline utilities/constant.py]: SAMPLE_SIZE - what exactly is this ?

Comment (Akhilleshgoswami) [reply utilities/constant.py]: it's mean the total number of block from where we will fetch  the gas fee

File: services/contracts/cruize/cruize_contract.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    def deposit_to_cruize(self, amount=None):
+        result = {"transaction_hash": None, "error": None}
+        amount = amount["amount"]
+
+        # need to pre calculate the price
+        # average apis call time 20s
```
Comment (Prithviraj8) [inline services/contracts/cruize/cruize_contract.py]: that's too high. Which api's call time are you referring to here ?

Comment (Akhilleshgoswami) [reply services/contracts/cruize/cruize_contract.py]: it happens when we fetch the gas fee from the most 10 (it may vary)recent  blocks

Comment (Akhilleshgoswami) [reply services/contracts/cruize/cruize_contract.py]: any api that will interact  with the chain and where we are using the gas strategy function to calculate the gas fee

File: utilities/constant.py (hunk 6 above)
Comment (Prithviraj8) [reply utilities/constant.py]: Please rename it to something more appropriate.

File: services/contracts/cruize/cruize_contract.py (hunk 7 above)
Comment (Prithviraj8) [reply services/contracts/cruize/cruize_contract.py]: where is deposit_to_cruize being called from ?
the deposit api in CruizeOperations viewset right ?

Comment (Prithviraj8) [reply services/contracts/cruize/cruize_contract.py]: no wait, I don't this being called from CruizeOperations Viewset. So where is this being used ? Not being used yet?

Comment (Akhilleshgoswami) [reply services/contracts/cruize/cruize_contract.py]: it's being called from the CruizeOperation view set ,

Comment (Akhilleshgoswami) [reply services/contracts/cruize/cruize_contract.py]: Currently, we don't have a smart contract deployed on the net that's why we are not using this function in CruizeOperation. basically, it's an estimated time. when I use the get strategy  function with an APIs call it takes time more than 20 sec to fetch the data  that's why I write this comment  that we need to think about this

File: services/contracts/load_contract.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+from decouple import config
+from web3 import Web3
+from web3.middleware import geth_poa_middleware
+
+<<<<<<< HEAD
+from services import DydxAdmin
+
+=======
+>>>>>>> ffd00aa0d5f93d480e6f0ef2786ab2937aeee7e1
```
Comment (Prithviraj8) [inline services/contracts/load_contract.py]: conflict

File: settings_config/celery_config.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    result_accept_content=["json"],
+    task_always_eager=True,
+    task_store_eager_result=True,
+    broker_url="redis://127.0.0.1:6379/0",
+    result_backend="redis://127.0.0.1:6379/1",
+)
+app.config_from_object(settings, namespace="CELERY")
+app.autodiscover_tasks(lambda: settings.INSTALLED_APPS)
+app.conf.beat_schedule = {
+    "borrow_from_aave": {"task": "borrow_from_aave", "schedule": 1.0},
```
Comment (Prithviraj8) [inline settings_config/celery_config.py]: Do we need this ?

File: settings_config/celery_config.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    "check_withdrawal": {"task": "check_withdrawal", "schedule": 1.0},
+    "deposit_to_gnosis": {"task": "deposit_to_gnosis", "schedule": 20.0},
```
Comment (Prithviraj8) [inline settings_config/celery_config.py]: remove this ?

Comment (Prithviraj8) [reply settings_config/celery_config.py]: Add the tasks for opening and closing no ?

File: services/contracts/cruize/cruize_contract.py (hunk 5 above)
Comment (Prithviraj8) [reply services/contracts/cruize/cruize_contract.py]: hey ??

File: dydx_user/views.py [hunk 11]
```diff
@@ -38,3 +34,15 @@ def register_user(self, request):
         except Exception as e:
             result["error"] = str(e)
             return Response(result, status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+    def deposit_test_fund(self, request):
+        result = {"message": None, "error": None}
+        admin = DydxAdmin()
+        try:
+            fund_detiels = admin.deposit_fund()
+            result["message"] = fund_detiels["data"]["transfer"]
```
Comment (Prithviraj8) [inline dydx_user/views.py]: fund_detiels -> fund_details

File: cruize_operations/views.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+            return Response(result, status.HTTP_200_OK)
+        except Exception as e:
+            return Response(e, status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+    def deposit(self, request):
+        self.serializer_class = AmountSerializers
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        amount = serializer.data
+        cruize_ref = Cruize()
```
Comment (Prithviraj8) [inline cruize_operations/views.py]: cruize_ref -> cruize_contract_ref

Comment (Prithviraj8) [reply cruize_operations/views.py]: Do this for any other contract's ref you're using.

Comment (Akhilleshgoswami) [reply cruize_operations/views.py]: done

File: dydx_operations/views.py [hunk 13]
```diff
@@ -26,8 +28,8 @@ def slow_withdrawal(self, request):
         data = serializer.data
         try:
             db_ref = OrderManager()
-            withdarwal_ref = DydxWithdrawal()
-            withdrawal_data = withdarwal_ref.slow_withdrawal(data)
+            withdrawal_ref = DydxWithdrawal()
```
Comment (Prithviraj8) [inline dydx_operations/views.py]: withdrawal_ref -> dydx_withdrawal_obj

Name class object variables the same as the ClassNames

File: dydx_user/views.py [hunk 14]
```diff
@@ -38,3 +34,15 @@ def register_user(self, request):
...
+        result = {"message": None, "error": None}
+        admin = DydxAdmin()
+        try:
+            fund_details = admin.deposit_fund()
+            result["message"] = fund_details["data"]["transfer"]
+            return Response(result, status.HTTP_200_OK)
+        except Exception as e:
+            e = vars(e)
+            result["error"] = e["msg"]["errors"][0]["msg"]
+            return Response(result, status.HTTP_500_INTERNAL_SERVER_ERROR)
```
Comment (Prithviraj8) [inline dydx_user/views.py]: why did we add this in dydx_user ? 
shouldn't it go to dydx_operations ?

File: dydx_user/urls.py [hunk 15]
```diff
@@ -5,4 +5,5 @@
 urlpatterns = [
     path("position_id", User.as_view({"get": "position_id"})),
     path("register", User.as_view({"post": "register_user"})),
+    path("deposit/test", User.as_view({"post": "deposit_test_fund"})),
```
Comment (Prithviraj8) [inline dydx_user/urls.py]: The api should go to dydx_operations

File: dydx_operations/views.py (hunk 2 above)
Comment (Prithviraj8) [reply dydx_operations/views.py:65]: Confirming, Do we need this?

File: utilities/constant.py [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+
+# price params
+MAX_WAIT_SECONDS = 60
+BLOCK_SAMPLE_SIZE = 20
+PROBABILITY = 100
+RINKEBY_CHAIN_ID = 4
+MAINNET_CHAIN_ID = 1
+
+# DYDX CONSTANT
+SIGNATURE = "x0"
```
Comment (Prithviraj8) [inline utilities/constant.py]: this should be 0x

**Review by Prithviraj8 — APPROVED** [review]
LGTM

**Pull request merged**